src/HostScene.cpp

```diff
@@ -0,0 +1,246 @@
+#include "HostScene.h"
+#include <glm/glm.hpp>
+#include <thread>
+#include <set>
+
+#define STB_IMAGE_IMPLEMENTATION
+#include "../third-party/stb_image.h"
+
+
+unsigned char floatToByte(float value) {
+    if (value <= 0.0)
+        return 0;
+    if (1.0 <= value)
+        return 255;
+    return static_cast<unsigned char>(256.0 * value);
+}
+
+HostTexture SceneLoader::loadTextureFromFile(const std::string& filename) {
+    float *fdata = nullptr; 
+    int bytes_per_pixel = 3;
+    auto n = bytes_per_pixel; 
+    std::vector<float3> data;
+    int width, height;
+
+    fdata = stbi_loadf(filename.c_str(), &width, &height, &n, bytes_per_pixel);
+    if (fdata == nullptr) {
+        throw std::runtime_error("Cannot load texture data, path: " + filename);
+    }
+
+    data.resize(width * height);
+    int total_bytes = width * height * bytes_per_pixel;
+
+    for(int i = 0, j = 0; i < total_bytes; i += 3, j++) {
+        data[j] = make_float3(
+            floatToByte(fdata[i]),
+            floatToByte(fdata[i + 1]),
+            floatToByte(fdata[i + 2])
+        );
+    }
+
+    STBI_FREE(fdata);
+    HostTexture tex{width, height, std::move(data)};
+    return tex;
+}
+
+std::vector<HostTexture> SceneLoader::loadTextures(const aiScene *scene, const std::string &resDir) {
+    std::vector<HostTexture> textures;
+    for (int materialIdx = 0; materialIdx < scene->mNumMaterials; materialIdx++) {
+        aiMaterial *material = scene->mMaterials[materialIdx];
+        for (int texType = aiTextureType_NONE; texType <= AI_TEXTURE_TYPE_MAX; texType++) {
+            auto textureType = static_cast<aiTextureType>(texType);
+            size_t texCnt = material->GetTextureCount(textureType);
+            for (size_t i = 0; i < texCnt; i++) {
+                aiTextureMapMode texMapMode[2]; 
+                aiString textPath;
+                aiReturn retStatus = material->GetTexture(textureType, i, &textPath, nullptr, nullptr, nullptr, nullptr, &texMapMode[0]);
+                if (retStatus != aiReturn_SUCCESS || textPath.length == 0) {
+                    continue;
+                }
+
+                std::string path = resDir + "/" + textPath.C_Str();
+                textureDataCache_[path] = textures.size();
+                textures.push_back(loadTextureFromFile(path));
+            }
+        }
+    }
+    return textures;
+}
+
+glm::mat4 convertMatrix(const aiMatrix4x4 &m) {
+    glm::mat4 ret;
+    for (int i = 0; i < 4; i++) {
+        for (int j = 0; j < 4; j++) {
+            ret[j][i] = m[i][j];
+        }
+    }
+    return ret;
+}
+
```

Comment:
Looks like a regression - cannot we just use glm::mat4 all the time? I guess it's not used during the path-tracing so we might skip this one.

Reply:
I managed to skip this, we dont need to do transforms ourselves - assimp is awesome